### ISSUE TITLE
RC Rolling Update does not wait for new pods to be ready and healthy before killing old ones

### ISSUE THREAD

**Issue by azr** [kubernetes/kubernetes#15557]
Hello there :)

I have a pod that takes nearly 20 minutes before it can serve anything.

It's readiness and liveness probes are correctly done.
Problem is when I run a rolling update, the old pods gets killed after one minute no matter what.
So, as it takes 20m for the new ones to get ready we have to wait 19m to serve anything.

Here is my awesome bash proof :
```
~/hb [master]
$ kubectl get pods
NAME                                                READY     STATUS    RESTARTS   AGE
hb-6d1ca57cd64328988bbaa26e58edf90737f368f3-w2pma   1/1       Running   0          31m
hb-703b014e8bf330b0beb05d692d1c71a9977323bb-2eiri   0/1       Running   0          1m
hb-703b014e8bf330b0beb05d692d1c71a9977323bb-9snv3   0/1       Running   0          2m
~/hb [master]
$ kubectl get pods
NAME                                                READY     STATUS    RESTARTS   AGE
hb-703b014e8bf330b0beb05d692d1c71a9977323bb-2eiri   0/1       Running   0          1m
hb-703b014e8bf330b0beb05d692d1c71a9977323bb-9snv3   0/1       Running   0          2m
~/hb [master]
$ kubectl get pods
NAME                                                READY     STATUS    RESTARTS   AGE
hb-703b014e8bf330b0beb05d692d1c71a9977323bb-2eiri   0/1       Running   0          14m
hb-703b014e8bf330b0beb05d692d1c71a9977323bb-9snv3   0/1       Running   0          16m
```
...leading to "booes this service sux"

Also note that those 20min will grow more and more with our datas, so the *--update-period flag* won't do it, plus it's too manual :D.

I there any way I can help fixing this issue ? I have time to work into this.

Cheers !

**Comment by derekwaynecarr** [comment]
This feels like a bug.

/cc @ironcladlou

**Comment by nikhiljindal** [comment]
@azr Are you using HEAD or release-1.0? This should work fine on HEAD.
Can you give us the output of ```kubectl version```.

**Comment by ironcladlou** [comment]
@azr, what timeout value are you specifying for the rolling update?

**Comment by azr** [comment]
@ironcladlou I am not using timeouts values at all, I believe the default is one minute. As said up there, this is not a solution for us, the boot time is going do grow with time, we'd prefer the new revision to tell when it's ready.

@nikhiljindal I'm using google cloud's kubernetes :
```
Client Version: version.Info{Major:"1", Minor:"0", GitVersion:"v1.0.6", GitCommit:"388061f00f0d9e4d641f9ed4971c775e1654579d", GitTreeState:"clean"}
Server Version: version.Info{Major:"1", Minor:"0", GitVersion:"v1.0.6", GitCommit:"388061f00f0d9e4d641f9ed4971c775e1654579d", GitTreeState:"clean"}
```

**Comment by ironcladlou** [comment]
@azr There's not currently any support for "indefinite" timeouts, so you're probably being affected by the default (it's the timeout value which will cause the readiness checks to fail, not the update period).

I think for now your best bet is to just set some incredibly large timeout (like 24 hours). @nikhiljindal, do you feel we should support indefinite timeouts?

**Comment by azr** [comment]
@ironcladlou are you saying that if I set 24h as a timeout, then the old one will be killed only once the new one is ready ?

**Comment by azr** [comment]
While I'm here, I'm pretty sure you should support indefinite rolling update, I've seen many production sql services take more that one hour to boot.

**Comment by ironcladlou** [comment]
@azr 
>@ironcladlou are you saying that if I set 24h as a timeout, then the old pod will be killed only once the new one is ready ?

I could have misunderstood your original problem. The readiness checks affect how the rolling updater makes progress: the updater counts the number of "ready" pods in order to decide whether scale-down can occur (e.g., if configuration dictates that the available pod count should never fall below 10, there must be 10 "ready" before the updater will scale down). If your pods take longer to become ready than can be observed given the timeout (i.e. the pods always appear unready), I'd expect the updater to stall awaiting pod readiness, and eventually just fail because it was unable to make progress.

Are you saying that the updater is scaling down pods which aren't ready (as defined by readiness checks)?

**Comment by azr** [comment]
@ironcladlou ,  yes ! I tried to demonstrate that in my console debug output, the old pods got killed before any of the new pod were ready.

**Comment by ironcladlou** [comment]
@azr If you're able to reproduce the behavior, could you please share the YAML of one of the pods which was scaled down prematurely, preferably as soon as possible after it's scaled down?

**Comment by azr** [comment]
Ah you asked yaml, sorry I did json, but anyways here it is :

With two large enougth nodes in my cluster, I had two zsh loop over `kubectl get pods --output=json` forever and for both of them the last time when I saw my old pod it this one :

My go program received a signal terminated when I check the logs & with `docker ps -a`

```json
{
  "kind": "List",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "items": [
    {
      "kind": "Pod", // new pod
      "apiVersion": "v1",
      "metadata": {
        "name": "hb-03d43-5b6dm",
        "generateName": "hb-03d43-",
        "namespace": "default",
        "selfLink": "/api/v1/namespaces/default/pods/hb-03d43-5b6dm",
        "uid": "d02bbd5f-727e-11e5-89f7-42010af0004b",
        "resourceVersion": "69265",
        "creationTimestamp": "2015-10-14T14:21:08Z",
        "labels": {
          "app": "hb",
          "deployment": "03d43",
          "state": "serving"
        },
        "annotations": {
          "kubernetes.io/created-by": "{\"kind\":\"SerializedReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"ReplicationController\",\"namespace\":\"default\",\"name\":\"hb-03d43\",\"uid\":\"ce563f24-727e-11e5-89f7-42010af0004b\",\"apiVersion\":\"v1\",\"resourceVersion\":\"69236\"}}"
        }
      },
      "spec": {
        "volumes": [
          {
            "name": "default-token-ypka4",
            "secret": {
              "secretName": "default-token-ypka4"
            }
          }
        ],
        "containers": [
          {
            "name": "hb-kubernetes",
            "image": "gcr.io/my_project/hb:03d43",
            "command": [
              ...
            ],
            "ports": [
              {
                "name": "http-stats",
                "containerPort": 8080,
                "protocol": "TCP"
              }
            ],
            "resources": {
              "limits": {
                "cpu": "100m"
              }
            },
            "volumeMounts": [
              {
                "name": "default-token-ypka4",
                "readOnly": true,
                "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount"
              }
            ],
            "livenessProbe": {
              "tcpSocket": {
                "port": 8080
              },
              "timeoutSeconds": 1
            },
            "readinessProbe": {
              "httpGet": {
                "path": "/",
                "port": 8080,
                "scheme": "HTTP"
              },
              "timeoutSeconds": 1
            },
            "terminationMessagePath": "/dev/termination-log",
            "imagePullPolicy": "Always"
          }
        ],
        "restartPolicy": "Always",
        "dnsPolicy": "ClusterFirst",
        "serviceAccountName": "default",
        "serviceAccount": "default",
        "nodeName": "gke-hb-e0cddd12-node-opk0"
      },
      "status": {
        "phase": "Running",
        "conditions": [
          {
            "type": "Ready",
            "status": "False"
          }
        ],
        "hostIP": "10.240.0.5",
        "podIP": "10.220.1.3",
        "startTime": "2015-10-14T14:21:08Z",
        "containerStatuses": [
          {
            "name": "hb-kubernetes",
            "state": {
              "running": {
                "startedAt": "2015-10-14T14:21:39Z"
              }
            },
            "lastState": {
              "terminated": {
                "exitCode": 2,
                "startedAt": "2015-10-14T14:21:38Z",
                "finishedAt": "2015-10-14T14:21:38Z",
                "containerID": "docker://0ec790f2a1c4159cc116fe99c3765a724493caae4d188ca2166c4ad7744c20a0"
              }
            },
            "ready": false,
            "restartCount": 1,
            "image": "gcr.io/my_project/hb:03d43",
            "imageID": "docker://b48c358fc181dd0088f52efcb84fa25ffc220bd5b51a7f98741b8a0b59eb36c9",
            "containerID": "docker://5b6a6ae6dbf7b11d749d41b3c489b4907ec9489e167ebfa7c3927ed28027dfc7"
          }
        ]
      }
    },

    {
      "kind": "Pod",  // old pod
      "apiVersion": "v1",
      "metadata": {
        "name": "hb-a1bec-1qpom",
        "generateName": "hb-a1bec-",
        "namespace": "default",
        "selfLink": "/api/v1/namespaces/default/pods/hb-a1bec-1qpom",
        "uid": "4536fe35-7271-11e5-89f7-42010af0004b",
        "resourceVersion": "68659",
        "creationTimestamp": "2015-10-14T12:44:12Z",
        "labels": {
          "app": "hb",
          "deployment": "a1bec",
          "state": "serving"
        },
        "annotations": {
          "kubernetes.io/created-by": "{\"kind\":\"SerializedReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"ReplicationController\",\"namespace\":\"default\",\"name\":\"hb-a1bec\",\"uid\":\"4361af0d-7271-11e5-89f7-42010af0004b\",\"apiVersion\":\"v1\",\"resourceVersion\":\"68358\"}}"
        }
      },
      "spec": {
        "volumes": [
          {
            "name": "default-token-ypka4",
            "secret": {
              "secretName": "default-token-ypka4"
            }
          }
        ],
        "containers": [
          {
            "name": "hb-kubernetes",
            "image": "gcr.io/my_project/hb:a1bec",
            "command": [
              ...
            ],
            "ports": [
              {
                "name": "http-stats",
                "containerPort": 8080,
                "protocol": "TCP"
              }
            ],
            "resources": {
              "limits": {
                "cpu": "100m"
              }
            },
            "volumeMounts": [
              {
                "name": "default-token-ypka4",
                "readOnly": true,
                "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount"
              }
            ],
            "livenessProbe": {
              "tcpSocket": {
                "port": 8080
              },
              "timeoutSeconds": 1
            },
            "readinessProbe": {
              "httpGet": {
                "path": "/",
                "port": 8080,
                "scheme": "HTTP"
              },
              "timeoutSeconds": 1
            },
            "terminationMessagePath": "/dev/termination-log",
            "imagePullPolicy": "Always"
          }
        ],
        "restartPolicy": "Always",
        "dnsPolicy": "ClusterFirst",
        "serviceAccountName": "default",
        "serviceAccount": "default",
        "nodeName": "gke-hb-e0cddd12-node-gn6n"
      },
      "status": {
        "phase": "Running",
        "conditions": [
          {
            "type": "Ready",
            "status": "True"
          }
        ],
        "hostIP": "10.240.0.6",
        "podIP": "10.220.0.11",
        "startTime": "2015-10-14T12:44:12Z",
        "containerStatuses": [
          {
            "name": "hb-kubernetes",
            "state": {
              "running": {
                "startedAt": "2015-10-14T12:44:34Z"
              }
            },
            "lastState": {
              
            },
            "ready": true,
            "restartCount": 0,
            "image": "gcr.io/my_project/hb:a1bec",
            "imageID": "docker://aa306453ae54f7ccad1fad93c3a14e674ac7b9f3dda3bf909672d15cffb7162f",
            "containerID": "docker://d94c7875259709bc6bb2e920b88353ec2f3051b8c089c7e8d08b85fbae35aa8d"
          }
        ]
      }
    }
  ]
}
```

**Comment by azr** [comment]
the same happens with replication set to 2

**Comment by nikhiljindal** [comment]
@ironcladlou Before you go further debugging this, I would like to point out that @azr is using v1.0.6 on GKE.
Isnt it correct to say that what @azr is experiencing is expected for that version?
We have now fixed the problem on HEAD, so that we wait for the new pods to become ready before we scale down old pods, but that has not been part of any release it.

If I am correct, @azr this should be fixed for you in the next release (v1.1). As @ironcladlou pointed out, you will have to set the timeout to a large value.

**Comment by ironcladlou** [comment]
@nikhiljindal is right, thanks.

**Comment by nikhiljindal** [comment]
Re: @azr's questions about GKE:
> Should I move to HEAD ? Can I ? Wil google update it's version ? How do I know when they will ? ( > sorry that's a lot of questions ! :) )

Only release versions are supported on GKE. We are planning to release kubernetes v1.1 in early November and your cluster on GKE will be automatically updated to that version soon after.
cc @a-robinson and @roberthbailey incase they want to add more.

**Comment by a-robinson** [comment]
Yup, GKE should be up-to-date with the 1.1 release within about a week after the release finalized.

**Comment by azr** [comment]
All right then ! I'll set a huge time until 1.1 is released :)

Thank you all guys for your awesome support !

**Comment by azr** [comment]
By the way I just tried the rolling update with `--timeout="-22m0s", the old one still got killed after a minute.

**Comment by azr** [comment]
If I read the doc it says
```
$ kubectl rolling-update -h
      ...
      --timeout="5m0s": Max time to wait for a replication controller to update before giving up. Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
      --update-period="1m0s": Time to wait between updating pods. Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
   ...
```
That probably means that I need to have 2 replicas and use the `--update-period` option, instead of ```--timeout``` to achieve zero downtime.

**Comment by nikhiljindal** [comment]
Yes, you need to set both update-period and timeout to large values.
